// @sudoku/sencode/index.js

import { GRID_COORDS, GRID_LENGTH, SENCODE_REGEX, SENCODE_REGEX_DIRECT, SENCODE_SEPARATOR as SEPARATOR, SENCODE_SEPARATOR_REVERSE as SEPARATOR_REVERSE } from '@sudoku/constants';
import Base62 from './base62';

/**
 * @param {number[]} flatGrid
 * @returns {boolean}
 */
function shouldReverse(flatGrid) {
	for (let cell = 0; cell < GRID_LENGTH; cell++) {
		if (flatGrid[(GRID_LENGTH - 1) - cell] !== 0) {
			return false;
		} else if (flatGrid[cell] !== 0) {
			return true;
		}
	}

	return false;
}


/**
 * @param {number[][]} sudoku
 * @returns {string}
 */
export function encodeSudoku(sudoku) {
	/** @type number[] */
	const flatGrid = sudoku.flat();

	const reversed = shouldReverse(flatGrid);
	if (reversed) {
		// Reverse sudoku array
		flatGrid.reverse();
	}

	let structure = '';
	let numbers = '';

	for (let cell = 0; cell < GRID_LENGTH; cell++) {
		structure += (flatGrid[cell] === 0 ? '0' : '1');

		if (flatGrid[cell] > 0) {
			numbers += flatGrid[cell] - 1; // Subtract 1 so the number gets smaller
		}
	}

	return Base62.encode(BigInt('0b' + structure)) +
	       (reversed ? SEPARATOR_REVERSE : SEPARATOR) +
	       Base62.encode(BigInt(numbers));
}

export function decodeSencodeDirect(sencode) {
	// example input: https://www.sudokuwiki.org/sudoku.htm?bd=100400006046091080005020000000500109090000050402009000000010900080930560500008004
	// examplt output: 100400006046091080005020000000500109090000050402009000000010900080930560500008004

	// 提取URL中最后的数独字符串
	const sudokuString = sencode.split('=')[1];

	// 初始化一个9x9的二维数组
	const grid = Array.from({ length: 9 }, () => Array(9).fill(0));

	// 将数独字符串转换为9x9的二维数组
	for (let i = 0; i < sudokuString.length; i++) {
		const row = Math.floor(i / 9);
		const col = i % 9;
		grid[row][col] = parseInt(sudokuString[i], 10);
	}

	return grid;
}




/**
 * @param {string} sencode
 * @returns {number[][]}
 */
export function decodeSencode(sencode) {
	let grid = [
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
		[0, 0, 0, 0, 0, 0, 0, 0, 0],
	];

	const reversed = sencode.indexOf(SEPARATOR_REVERSE) !== -1;

	const [encodedStructure, encodedNumbers] = sencode.split(reversed ? SEPARATOR_REVERSE : SEPARATOR);

	const structure = Base62.decode(encodedStructure)
	                        .toString(2)
	                        .padStart(GRID_LENGTH, '0');

	let numberCount = 0;
	for (let cell = 0; cell < GRID_LENGTH; cell++) {
		if (structure[cell] === '1') {
			numberCount++;
		}
	}

	let numbers = Base62.decode(encodedNumbers)
	                    .toString()
	                    .padStart(numberCount, '0')
	                    .split('');

	for (let cell = 0; cell < GRID_LENGTH; cell++) {
		if (structure[cell] === '1') {
			const [row, col] = GRID_COORDS[cell];
			grid[row][col] = numbers.shift() * 1 + 1;
		}
	}

	if (reversed) {
		// Reverse Grid
		grid = grid.reverse().map(row => row.reverse());
	}

	return grid;
}

/**
 * @param {string} sencode
 * @returns {boolean}
 */
export function validateSencode(sencode) {
	return sencode && sencode.trim().length !== 0 && SENCODE_REGEX.test(sencode);
}

/**
 * @param {string} sencode
 * @returns {boolean}
 */
export function validateSencodeDirect(sencode) {
	return sencode && sencode.trim().length !== 0 && SENCODE_REGEX_DIRECT.test(sencode);
}
